### PR TITLE
Devenv: Add a dashboard with all panels

### PR DESCRIPTION
**Pull request by tskarhed**
**What this PR does / why we need it**:

This was created to easily do accessibility testing on as many panels as possible at once.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Just go into `devenv`and run `./setup.sh`. Then see if you find the `Panel tests - All panels`dashboard.

### CODE REVIEW

**Comment by ryantxu** [comment]
@sdboyer this is one we should look at for two reasons:
1. get it to pass
2. should we have a test that can assert this dashboard actually has all builtin panel plugins?
3. our validation should be very forgiving for panel types that are not defined

**Comment by sdboyer** [comment]
> should we have a test that can assert this dashboard actually has all builtin panel plugins?

yes. we want CI-produced certainty that our guarantees extend to all builtin/dist panel plugins.

or, if you prefer: `∀p∈P: V(p)` :ducks:

> our validation should be very forgiving for panel types that are not defined

Yeah, we need a clear strategy here. Short-term, what's causing the problem is [this conundrum](https://github.com/grafana/grafana/blob/b88f8bacf3aa258c33722dd894f40ac762382ff5/cue/data/gen.cue#L286-L291).

**Review by hugohaggmark — APPROVED** [review]
LGTM 🎉  but seems the build is failing because of schema errors.

**Comment by sdboyer** [comment]
yep, I have to finish up #38727, then this will be unblocked

**Comment by sdboyer** [comment]
Error on that PR is currently because nulls need to be stripped (run `scripts/stripnulls.sh`) - i need to write a guide to this, sorry (#38712)

**Review by thisisobate — CHANGES_REQUESTED** [review]
Need to create dashboard json that doesn't violate the schema.

File: devenv/dev-dashboards/all-panels.json
Comment (sdboyer) [inline devenv/dev-dashboards/all-panels.json:541]: This looks like a proper schema violation - heatmap isn't schematized, but this is within the segment of the structure that's owned by the core schema. Either we need to add something for `cards`, or refactor the way that heatmap emits data.

i'm also quite concerned that there are several other values in inappropriate places, here - we made a hack for the old Graph panel, but i really don't want to do another. Ideally, we'd refactor to ensure that heatmap stays within the boundaries of where panel config/opts are supposed to reside.

File: devenv/dev-dashboards/all-panels.json
Comment (sdboyer) [inline devenv/dev-dashboards/all-panels.json:558]: All three of these will be a schema violation, even though (ugh) they're not showing up in test results.

File: devenv/dev-dashboards/all-panels.json
Comment (sdboyer) [inline devenv/dev-dashboards/all-panels.json:563]: this violates schema

File: devenv/dev-dashboards/all-panels.json
Comment (sdboyer) [inline devenv/dev-dashboards/all-panels.json:574]: i believe this is also a violation

**Review by sdboyer — CHANGES_REQUESTED** [review]
Unfortunately, we do have a bunch of legit schema violations, all of which will need to be addressed before we can merge this.

REALLY unfortunately, it would appear that a side effect of my refactor in #38727 is to make error messages even _more_ obscure. Now, for whatever reason, they're not linked back to the models.cue file from the original plugin - instead it's just a big long list of file positions within the composition logic in dashboard.cue 😠 

For now, we'll just have to live with that, and deal with the issues directly here as they arise.

File: devenv/dev-dashboards/all-panels.json
Comment (torkelo) [reply devenv/dev-dashboards/all-panels.json:541]: this is like the old graph panel, it adds all its options to the base panel model

File: devenv/dev-dashboards/all-panels.json
Comment (torkelo) [reply devenv/dev-dashboards/all-panels.json:574]: these are all valid heatmap panel options, can we exclude panels with type: "heatmap" or we will have to remove this panel from this dashboard

**Comment by torkelo** [comment]
I do not think we shoud close this, we need to figure out how we can relax the schema validation for old panel plugins like heatmap, or remove the heatmap panel from this dashboard if we cannot use heatmap in devenv dashboards without triggering schema valdiation. Any ideas for how to proceed @sdboyer ?

File: devenv/dev-dashboards/all-panels.json
Comment (sdboyer) [reply devenv/dev-dashboards/all-panels.json:541]: OK. There's two obvious paths here:

1. If we can, as with the old graph panel, promise that the heatmap is deprecated and its schema will never change again, then we can safely (if still inelegantly) add it to the core dashboard schema, same as with #GraphPanel
2. If we can't make that promise, it implies we'd need to refactor heatmap to follow the same rules as other panel plugins, define a `models.cue` for it and confine any of its unique keys and values to the `options` and `fieldConfig.defaults` subtrees. We'd also need to write a TS migration that transforms existing heatmap panels into the new, rigorously-separated form in order for the schema system to be fully capable of handling it.

We also can combine these prior to GA of the schema system - embed a `#HeatmapPanel` definition in the core dashboard scuemata for now in order to get validation passing, then remove it once we've refactored heatmap to follow rigorous separation rules.

Also, to get out ahead of it, i've also seen `table-old` seems to add its options to the base panel, too - are there any others?

**Comment by sdboyer** [comment]
OK. There's two obvious paths to dealing with heatmap here:

1. If we can, as with the old graph panel, promise that the heatmap is deprecated and its schema will never change again, then we can safely (if still inelegantly) add it to the core dashboard schema, same as with #GraphPanel
2. If we can't make that promise, it implies we'd need to refactor heatmap to follow the same rules as other panel plugins, define a `models.cue` for it and confine any of its unique keys and values to the `options` and `fieldConfig.defaults` subtrees. We'd also need to write a TS migration that transforms existing heatmap panels into the new, rigorously-separated form in order for the schema system to be fully capable of handling it.

We also can combine these prior to GA of the schema system - embed a `#HeatmapPanel` definition in the core dashboard scuemata for now in order to get validation passing, then remove it once we've refactored heatmap to follow rigorous separation rules.

Also, to get out ahead of it, i've also seen `table-old` seems to add its options to the base panel, too - are there any others?

To be totally clear, though, we have to have a strategy in place for heatmap (and any other panels that do this) prior to actually saying that the schema system is truly ready for general use - while we can artificially prevent failures in devenv by just choosing not to add a heatmap plugin, that's clearly not a strategy that will fly in the real world.

**Comment by sdboyer** [comment]
Seems like we need a general solution for this, actually - extend openness for unknown plugin types to not only the plugin-owned `options` and `fieldConfig.defaults` spaces, but the whole panel object. Maybe we can ratchet that further down later, but i have a couple-line change that may be able to take care of this.

File: packages/grafana-schema/src/scuemata/dashboard/dashboard.cue
Comment (sdboyer) [inline packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:323]: This makes the base panel structure open for unknown panel types. That means that schema on explicitly stated fields will still be enforced for known and unknown panel types, but unknown types ones can introduce fields wherever they want on the Panel structure.

That will allow legacy angular panels - like heatmap, but also including third party ones, for which we can obviously never write special exceptions - will be able to pass validation.

Could plausibly get rid of the special `#GraphPanel` in the dashboard scuemata, too, though i'll look at that in a follow-up.

**Comment by sdboyer** [comment]
Pushed up a couple commits that should cause this to pass CI. The big one is a one-liner that permits full-on openness for unknown panel types - see [this comment](https://github.com/grafana/grafana/pull/38113/files#r709375617).

The other ones are small tweaks. I got rid of a link, as [we simply haven't specified those yet](https://github.com/grafana/grafana/blob/2a4b08eab93c1c8a4e35ea25c8521c1663662488/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L188-L193), and there's no reason to block the larger goal of this issue on what is an orthogonal TODO.

The other one we plausibly should address in this PR before merging - it seems like `gauge` panels sometimes don't generate an `orientation`, but that's marked as required in the schema and has no default value. We should either change the schema to include a default, or change something on the TS side so that an appropriate value is always there. (Seems like having `auto` be the default is probably what we want...?)

File: packages/grafana-schema/src/scuemata/dashboard/dashboard.cue
Comment (hugohaggmark) [reply packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:323]: I like this idea 🚀

Comment (torkelo) [reply packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:323]: nice!

**Comment by torkelo** [comment]
Ok, this dashboard was quite broken, it had "${GDEV_TESTDATA}" in the data source prop so caused lots of runtime errors, must have been an exported dashboard (those cannot be used in provisioning as they require user manual import to select data source mapping). Fixed that and fixed the broken geomap panel that caused the dashboard to hang (it used a streaming query which hangs the geomap/grafana). 

Removed the orientation prop on gauge and found the cause for why this appears some times, this is not a valid option for gauge, but is added in the migration to gauge from a panel like bar gauge and stat (that do have orientation), will fix in separate PR

**Comment by torkelo** [comment]
here is the fix for the gauge orientation prop issue: https://github.com/grafana/grafana/pull/39292

**Comment by sdboyer** [comment]
> but the gauge cue file is wrong, it's extending ui.SingleStatBaseOptions which I guess includes orientation but gauge has no such option

i'm gonna defer to @ryantxu on this - i don't have the knowledge to have an opinion as to how those core reusable types _should_ be written, so i'd just be asking other folks about what the proper fix ought to be.

**Comment by torkelo** [comment]
@sdboyer I search for `ui.SingleStatBaseOptions` but could not find where this was defined

**Comment by sdboyer** [comment]
@torkelo it's [here](https://github.com/grafana/grafana/blob/main/packages/grafana-schema/src/schema/mudball.cue#L11-L15). It's just like a Go import - the `ui` is the identifier that namespaces object in the imported package.

Looks like `SingleStatBaseOptions` is also used in the `stat` panel's models.cue, so it may or may not be correct to change that shared declaration.

This isn't a super-hard question, it should be answered in the same way we'd do it in typescript where there's a shared interface used by two other ones - should it actually be shared?

**Comment by torkelo** [comment]
ok, adding back orientation prop then

**Comment by sdboyer** [comment]
That works, too, we can work out what exactly should be in those in a separate PR. (arguably, a good thing to be done in #39292)

**Review by sdboyer — APPROVED** [review]

**Comment by hugohaggmark** [comment]
Thank you @torkelo and @sdboyer for completing this PR ❤️

**Pull request merged**
